### ISSUE TITLE
have_posts() is an impure function!

### ISSUE THREAD

**Issue by szepeviktor** [szepeviktor/phpstan-wordpress#42]
`wordpress-override.stub`

```php
<?php

/**
 * Whether current WordPress query has results to loop over.
 *
 * @since 1.5.0
 *
 * @global WP_Query $wp_query Global WP_Query instance.
 *
 * @return bool
 * @phpstan-impure
 */
function have_posts()
{
}
```

**Comment by dingo-d** [comment]
Is this related to why the phpstan reports for this:

```php
if ($query->have_posts()) {
	while ($query->have_posts()) {
		$query->the_post();

		$list[] = [
			'id' => \get_the_ID(),
			'title' => \get_the_title(),
			'url' => \get_permalink(),
		];
	}
}
\wp_reset_postdata();
```

`While loop condition is always true.` ?

I got this notice in phpstan once I updated to phpstan v1

**Comment by dingo-d** [comment]
Thanks, I'll just have to ignore this line per line 😅

**Comment by jeherve** [comment]
> I'll just have to ignore this line per line

You could try to edit your `phpstan.neon.dist` to always ignore that warning, by targetting something like:
`message: '#While loop condition is always true#'`

While it may catch other use-cases, it may help a bit instead of doing it line by line?

**Comment by dingo-d** [comment]
Yeah, I'd like not to ignore all those cases, because it might ignore legitimate notices 🙂 So one by one would be good alternative 😄

**Comment by szepeviktor** [comment]
[wordpress-stubs.txt](https://github.com/szepeviktor/phpstan-wordpress/files/7722001/wordpress-stubs.txt)

Try with these stubs! It contains `class-wp-query.php` with all its source code.

**Comment by janw-me** [comment]
This isn't about WP_Query only if you wrap the `while` in the `if` it assumes it's true and always will untill the end of time.
The same happens with the ACF function `have_rows` used by the [repeater](https://www.advancedcustomfields.com/resources/repeater/) field.  
The example below gives an error 3 out of 4 times.

```php
<?php
// normal template
if ( have_posts() ) :
	while ( have_posts() ):
		the_post();
	endwhile;
else:
	echo 'fallback template';
endif;

// custom query
$query = new WP_Query();
if ( $query->have_posts() ) :
	while ( $query->have_posts() ):
		$query->the_post();
	endwhile;
else:
	echo 'fallback template';
endif;

// ACF > including https://github.com/php-stubs/acf-pro-stubs
if ( have_rows( 'featured' ) ):
	while ( have_rows( 'featured' ) ) :
		the_row();
	endwhile;
else:
	echo 'fallback template';
endif;

// Custom Query
// No if === no error
$query = new WP_Query();
while ( $query->have_posts() ):
	$query->the_post();
endwhile;
```

I'm willing to look into this although I have no experience with stanphp untill today...
If someone has pointers I would appreciate it ;)

**Comment by szepeviktor** [comment]
> I have no experience with stanphp untill today...

Hello Janw!
PHPStan (not stanphp) is an OOP tool, it won't work for meritorious impure function.
Please add `@phpstan-impure` to all stubs that need is. There is `sed` for you!

**Comment by herndlm** [comment]
Viktor did you handle this one already via 0c610fd or is this issue meant to be kept open for e.g. a more generic solution or something like that?

**Comment by szepeviktor** [comment]
I cannot answer. There are only wrong answers.
This is a pinned issue.
